### PR TITLE
Make raw memory access tunable

### PR DESCRIPTION
**Pull request by topimiettinen**
Modern systems shouldn't need direct access to raw memory
devices (/dev/mem, /dev/kmem, /dev/mergemem, dev/oldmem, /dev/port)
anymore, so let's make it tunable.

Signed-off-by: Topi Miettinen <toiwoton@gmail.com>

### CODE REVIEW

**Comment by topimiettinen** [comment]
This seems very straightforward, but there is a strange compile error:
```
Compiling refpolicy iscsi.mod module
m4 -E -E -D enable_ubac -D mls_num_sens=16 -D mls_num_cats=1024 -D mcs_num_cats=1024 -D hide_broken_symptoms -s support/divert.m4 policy/support/file_patterns.spt policy/support/ipc_patterns.spt policy/support/obj_perm_sets.spt policy/support/misc_patterns.spt policy/support/misc_macros.spt policy/support/mls_mcs_macros.spt policy/support/loadable_module.spt support/undivert.m4 tmp/generated_definitions.conf tmp/all_interfaces.conf policy/modules/system/iscsi.te support/fatal_error.m4 > tmp/iscsi.tmp
/usr/bin/checkmodule -m tmp/iscsi.tmp -o tmp/iscsi.mod
policy/modules/system/iscsi.te:87:ERROR 'syntax error' at token 'typeattribute' on line 3715:
#line 87
        typeattribute iscsid_t memory_raw_read;
/usr/bin/checkmodule:  error(s) encountered while parsing configuration
make: *** [Rules.modular:78: tmp/iscsi.mod] Error 1
```
I think the syntax in iscsi.te is same as everywhere else, so what could be the problem?

**Comment by doverride** [comment]
Topi Miettinen <notifications@github.com> writes:

> This seems very straightforward, but there is a strange compile error:
>
> Compiling refpolicy iscsi.mod module
> m4 -E -E -D enable_ubac -D mls_num_sens=16 -D mls_num_cats=1024 -D mcs_num_cats=1024 -D hide_broken_symptoms -s support/divert.m4 policy/support/file_patterns.spt policy/support/ipc_patterns.spt policy/support/obj_perm_sets.spt policy/support/misc_patterns.spt policy/support/misc_macros.spt policy/support/mls_mcs_macros.spt policy/support/loadable_module.spt support/undivert.m4 tmp/generated_definitions.conf tmp/all_interfaces.conf policy/modules/system/iscsi.te support/fatal_error.m4 > tmp/iscsi.tmp
> /usr/bin/checkmodule -m tmp/iscsi.tmp -o tmp/iscsi.mod
> policy/modules/system/iscsi.te:87:ERROR 'syntax error' at token 'typeattribute' on line 3715:
> #line 87
>         typeattribute iscsid_t memory_raw_read;
> /usr/bin/checkmodule:  error(s) encountered while parsing configuration
> make: *** [Rules.modular:78: tmp/iscsi.mod] Error 1
>
> I think the syntax in iscsi.te is same as everywhere else, so what could be the problem?

Limitation: you cannot associate with type attributes in conditionals.
just make the neverallow assertion type attribute associations unconditional and
instead only make the TE allow rules conditional.

>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

-- 
Key fingerprint = FCD2 3660 5D6B 9D27 7FC6  E0FF DA7E 521F 10F6 4098
https://sks-keyservers.net/pks/lookup?op=get&search=0xDA7E521F10F64098
Dominick Grift

**Comment by topimiettinen** [comment]
Thanks, that solves the problem.

**Comment by topimiettinen** [comment]
Should the boolean default to true (for compatibility) or false?

**Comment by pebenito** [comment]
If this access isn't needed anymore, we should consider removing the access instead. That would best be proposed to the mail list.

File: policy/support/obj_perm_sets.spt [hunk 1]
```diff
@@ -170,6 +170,7 @@ define(`create_file_perms',`{ getattr create open }')
 define(`rename_file_perms',`{ getattr rename }')
 define(`delete_file_perms',`{ getattr unlink }')
 define(`manage_file_perms',`{ create open getattr setattr read write append rename link unlink ioctl lock }')
+define(`manage_metadata_file_perms',`{ create getattr setattr rename link unlink }')
```
Comment (pebenito) [inline policy/support/obj_perm_sets.spt]: NAK. I don't think "metadata" fits in the name, here or in any of the other permission set and pattern additions.

File: policy/global_tunables [hunk 2]
```diff
@@ -32,6 +32,13 @@ gen_tunable(allow_execmod,false)
 ## </desc>
 gen_tunable(allow_execstack,false)
 
+## <desc>
+## <p>
+## Allow raw memory device (/dev/mem, /dev/kmem, /dev/mergemem, dev/oldmem, /dev/port) access for unconfined executables.
```
Comment (pebenito) [inline policy/global_tunables]: These domains are confined, not unconfined.

File: policy/modules/admin/readahead.te [hunk 3]
```diff
@@ -42,7 +42,6 @@ dev_getattr_generic_blk_files(readahead_t)
 dev_getattr_all_chr_files(readahead_t)
 dev_getattr_all_blk_files(readahead_t)
 dev_dontaudit_read_all_blk_files(readahead_t)
-dev_dontaudit_getattr_memory_dev(readahead_t)
```
Comment (pebenito) [inline policy/modules/admin/readahead.te]: I don't think this `dontaudit` should be removed.

File: policy/modules/services/abrt.te [hunk 4]
```diff
@@ -165,7 +165,6 @@ dev_getattr_all_blk_files(abrt_t)
 dev_read_rand(abrt_t)
 dev_read_urand(abrt_t)
 dev_rw_sysfs(abrt_t)
-dev_dontaudit_read_raw_memory(abrt_t)
```
Comment (pebenito) [inline policy/modules/services/abrt.te]: No need to remove.

File: policy/modules/system/mount.te [hunk 5]
```diff
@@ -77,7 +77,6 @@ dev_dontaudit_write_sysfs_dirs(mount_t)
 dev_rw_lvm_control(mount_t)
 dev_rw_loop_control(mount_t)
 dev_dontaudit_getattr_all_chr_files(mount_t)
-dev_dontaudit_getattr_memory_dev(mount_t)
```
Comment (pebenito) [inline policy/modules/system/mount.te]: Should not remove.

File: policy/modules/kernel/devices.if [hunk 6]
```diff
@@ -1420,13 +1448,45 @@ interface(`dev_manage_all_blk_files',`
...
+##	Domain allowed access.
+##	</summary>
+## </param>
+#
+interface(`dev_attr_memory_raw_write',`
+	gen_require(`
+		attribute memory_raw_write;
+	')
+	typeattribute $1 memory_raw_write;
 ')
```
Comment (pebenito) [inline policy/modules/kernel/devices.if]: See `storage_raw_read_fixed_disk_cond()` for an example of how to do this.

File: policy/support/obj_perm_sets.spt (hunk 1 above)
Comment (topimiettinen) [reply policy/support/obj_perm_sets.spt]: Removed

File: policy/global_tunables (hunk 2 above)
Comment (topimiettinen) [reply policy/global_tunables]: Fixed

File: policy/modules/admin/readahead.te (hunk 3 above)
Comment (topimiettinen) [reply policy/modules/admin/readahead.te]: Restored

File: policy/modules/services/abrt.te (hunk 4 above)
Comment (topimiettinen) [reply policy/modules/services/abrt.te]: Restored

File: policy/modules/system/mount.te (hunk 5 above)
Comment (topimiettinen) [reply policy/modules/system/mount.te]: Restored

File: policy/modules/kernel/devices.if (hunk 6 above)
Comment (topimiettinen) [reply policy/modules/kernel/devices.if]: Thanks, updated to match that.

**Comment by topimiettinen** [comment]
This still leaves unconditional access to kdumpctl_t, rmp_t, systemd_tmpfiles_t and udev_t through dev_manage_all_dev_nodes(). I didn't find an easy way to make that conditional, but perhaps that can be improved later.

File: policy/modules/kernel/devices.if [hunk 7]
```diff
@@ -2736,103 +2736,95 @@ interface(`dev_dontaudit_getattr_memory_dev',`
...
 ##	Domain allowed access.
 ##	</summary>
 ## </param>
+## <param name="tunable">
+##	<summary>
+##	Tunable to depend on
+##	</summary>
+## </param>
 #
-interface(`dev_read_raw_memory',`
```
Comment (pebenito) [inline policy/modules/kernel/devices.if]: All of the original interfaces must remain for backwards compatibility.

Comment (topimiettinen) [reply policy/modules/kernel/devices.if]: Ok, reverted

**Comment by pebenito** [comment]
Are you expecting any further changes? (is this still WIP?)

**Comment by topimiettinen** [comment]
No, unless someone has an idea how to restrict also kdumpctl_t, rmp_t, systemd_tmpfiles_t and udev_t.

**Pull request merged**